rclcpp_action/src/server.cpp

```diff
@@ -537,6 +537,11 @@ ServerBase::publish_result(const GoalUUID & uuid, std::shared_ptr<void> result_m
       }
     }
   }
+
+  // Publish a status message any time a goal handle changes state
+  publish_status();
+  // notify base so it can recalculate the expired goal timer
+  notify_goal_terminal_state();
```

Comment:
I think that the problem is lock order but not locking time window. so I believe with this fix, it still makes the deadlock problem. (in case, right after `rclcpp_action::ServerGoalHandle::succeed` is issued, then in another thread `rclcpp_action::ServerBase::execute_goal_request_received` is fired then calling user callback `handle_goal`)

I think `reentrant_mutex_` must be used internally except user callback.

Reply:
Yeah, I agree with this. This could solve only for my test case.

> I think reentrant_mutex_ must be used internally except user callback.
I changed the fix, is this what you thought?

Reply:
I think that this works, but i'd like to hear from others.